Mathlib/NumberTheory/FactorisationProperties.lean

```diff
@@ -0,0 +1,320 @@
+/-
+Copyright (c) 2024 Colin Jones. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Colin Jones
+-/
+import Mathlib.NumberTheory.Divisors
+import Mathlib.Tactic.NormNum.Prime
+import Mathlib.Algebra.GeomSum
+import Mathlib.Tactic.FinCases
+
+/-!
+## FactorisationProperties.lean
+This file defines abundant, semi-perfect, deficient, and weird numbers and formalizes their
```

Comment:
Follow-up question: does mathlib have perfect numbers already?
If not, the relation between Mersenne primes and even perfect numbers could be nice material to follow. Since this PR is long enough, a follow-up would be best, probably.

Reply:
Yes, I believe that would be an interesting direction to take this in the future.

Reply:
I'm also planning to add other proofs about these factorization properties later on.

Reply:
Mathlib does have perfect numbers, and I believe Aaron Anderson formalized them. Mersenne numbers were also formalized but by someone else.